File: lua/examples/guards/scripts/Player.lua
---@uuid c5671d19-9f1a-4286-8486-add4ebaadaec
---@class Player : Script
---@field sensitivity number
---@field camera Node
---@field power number
---@field bullet Prefab
---@field initial_bullet_velocity number
---@field shooting_range number
---@field reload_delay_sec number
---@field private reload_sec number
---@field private published boolean
---@field private collider Node
---@field private aim_y number
Player = script_class()


FRACTION_PLAYER = 0

---@param x number
function Player:turn(x)
    local rot_delta = Quaternion:from_axis_angle(Vector3.Y, self.sensitivity * x);

    self.node.local_rotation = self.node.local_rotation:mul_quat(rot_delta)
end

function Player:aim(y)
    self.aim_y = self.aim_y + y * self.sensitivity;

    self.aim_y = math.max(-math.pi / 2.0, math.min(self.aim_y, math.pi / 2.0))

    self.camera.local_rotation = Quaternion:from_axis_angle(
        Vector3.X,
        self.aim_y
    );
end

function Player:fire()
    local camera_pos = self.camera.global_position;

    local bullet_orientation = self.camera.global_rotation;

    Bullet:spawn(
        {
            prefab = self.bullet,
            origin = camera_pos,
            direction = bullet_orientation:mul_vec(Vector3.Z),
            initial_velocity = self.initial_bullet_velocity,
            author_collider = self.collider,
            range = self.shooting_range,
            fraction = FRACTION_PLAYER,
        }
    );
end


function Player:on_init()
    Window.cursor_grab = CursorGrabMode.Confined;

    self.collider = not_nil(self.node:find_collider_in_children(), "player collider missing")
end

function Player:on_start()
    self.node:subscribe_to()
end

function Player:on_message(message)
    if message.type == Bullet.HitMessage and message.fraction ~= FRACTION_PLAYER then
        Plugin:get("Game"):inc_wounds()
        print("player wounded!")
    end
end

function Player:on_update(dt)
    if self.reload_sec > 0.0 then
        self.reload_sec = self.reload_sec - dt;
    end
    if not self.published then
        self.published = true;
        Plugin:get("Game").player = self.node;
    end

    if Input:is_mouse_button_pressed(Input.MouseLeft) then
        if self.reload_sec <= 0.0 then
            self.reload_sec = self.reload_delay_sec;
            self:fire();
        end
    end

    local move_delta = Vector3.ZERO;

    if Input:is_key_pressed(KeyCode.W) then
        move_delta.z = move_delta.z + 1.0
    end
    if Input:is_key_pressed(KeyCode.S) then
        move_delta.z = move_delta.z - 1.0
    end
    if Input:is_key_pressed(KeyCode.A) then
        move_delta.x = move_delta.x + 1.0
    end
    if Input:is_key_pressed(KeyCode.D) then
        move_delta.x = move_delta.x - 1.0
    end

    self:turn(-Input.mouse_move.x)
    self:aim(Input.mouse_move.y)

    if move_delta:magnitude() > 0.001 then
        move_delta:normalize_inplace();
    end

    local self_rotation = self.node.local_rotation
    local move_delta = self_rotation:mul_vec(move_delta);
    local force = move_delta:mul(self.power);
    self.node:as_rigid_body():apply_force(force);
end

File: cs/examples/Guards/Player.cs
using FyroxLite;

[Uuid("c5671d19-9f1a-4286-8486-add4ebaadaec")]
public class Player: NodeScript
{
    private float sensitivity;
    private Node camera;
    private float power;
    private Prefab bullet;
    private float initial_bullet_velocity;
    private float shooting_range;
    private float reload_delay_sec;
    
    [HideInInspector]
    [Transient]
    private float ReloadSec;
    
    [HideInInspector]
    [Transient]
    private bool Published;
    
    [HideInInspector]
    [Transient]
    private Node Collider;
    
    [HideInInspector]
    [Transient]
    private float AimY;

    public const int FractionPlayer = 0;

    public void Turn(float x)
    {
        Quaternion rotDelta = Quaternion.FromEuler(Vector3.Up * sensitivity * x);
        Node.LocalRotation *= rotDelta;
    }

    public void Aim(float y)
    {
        AimY += y * sensitivity;
        AimY = Math.Max(-MathF.PI / 2.0f, Math.Min(AimY, MathF.PI / 2.0f));

        camera.LocalRotation = Quaternion.FromEuler(Vector3.Left * AimY);
    }

    public void Fire()
    {
        Vector3 cameraPos = camera.GlobalPosition;
        Quaternion bulletOrientation = camera.GlobalRotation;

        Bullet.Spawn(new Bullet.BulletSeed
        {
            Prefab = bullet,
            Origin = cameraPos,
            Direction = bulletOrientation * Vector3.Forward,
            InitialVelocity = initial_bullet_velocity,
            AuthorCollider = Collider,
            Range = shooting_range,
            Fraction = FractionPlayer,
        });
    }

    protected override void OnInit()
    {
        Window.CursorGrab = CursorGrabMode.Confined;
        Collider = Node.FindColliderInChildren() ?? throw new Exception("player collider missing");
    }

    protected override void OnStart()
    {
        Node.SubscribeTo();
    }

    protected override void OnMessage(object message)
    {
        if (message is BulletHitMessage hitMessage && hitMessage.Fraction != FractionPlayer)
        {
            Plugin.Get<Game>().IncWounds();
            Console.WriteLine("player wounded!");
        }
    }

    protected override void OnUpdate(float dt)
    {
        if (ReloadSec > 0.0f)
        {
            ReloadSec -= dt;
        }
        if (!Published)
        {
            Published = true;
            Plugin.Get<Game>().player = Node;
        }

        if (Input.IsMouseButton(Input.MouseLeft))
        {
            if (ReloadSec <= 0.0f)
            {
                ReloadSec = reload_delay_sec;
                Fire();
            }
        }

        Vector3 moveDelta = Vector3.Zero;

        if (Input.IsKey(KeyCode.W))
        {
            moveDelta.Z += 1.0f;
        }
        if (Input.IsKey(KeyCode.S))
        {
            moveDelta.Z -= 1.0f;
        }
        if (Input.IsKey(KeyCode.A))
        {
            moveDelta.X += 1.0f;
        }
        if (Input.IsKey(KeyCode.D))
        {
            moveDelta.X -= 1.0f;
        }

        Turn(-Input.MouseMove.X);
        Aim(Input.MouseMove.Y);

        if (moveDelta.Length() > 0.001f)
        {
            moveDelta = moveDelta.Normalized();
        }

        var selfRotation = Node.LocalRotation;
        Vector3 moveDirection = selfRotation * moveDelta;
        Vector3 force = moveDirection * power;
        Node.AsRigidBody().Value.ApplyForce(force);
    }
}
